File: backend/app/utils/heatmap_spread.py
"""Expand heatmap points for smooth leaflet.heat blending."""

from __future__ import annotations

from app.schemas.heatmap import HeatmapPoint

# (lat multiplier, lng multiplier, intensity factor)
_SPREAD_PATTERN: list[tuple[float, float, float]] = [
    (0.0, 0.0, 1.0),
    (1.0, 0.0, 0.92),
    (-1.0, 0.0, 0.88),
    (0.0, 1.0, 0.85),
    (0.0, -1.0, 0.8),
    (1.0, 1.0, 0.78),
    (1.0, -1.0, 0.76),
    (-1.0, 1.0, 0.74),
    (-1.0, -1.0, 0.72),
    (2.0, 0.0, 0.65),
    (-2.0, 0.0, 0.62),
    (0.0, 2.0, 0.6),
    (0.0, -2.0, 0.58),
    (1.5, 1.5, 0.55),
    (-1.5, 1.5, 0.53),
    (1.5, -1.5, 0.51),
    (-1.5, -1.5, 0.49),
]


def spread_step_degrees(
    north: float, south: float, east: float, west: float, grid: int, zoom: int
) -> float:
    lat_span = max(0.01, north - south)
    lon_span = max(0.01, east - west)
    base = max(lat_span, lon_span) / max(grid * 1.25, 8)
    zoom_factor = 1.0 + max(0, 14 - zoom) * 0.08
    return base * zoom_factor


def expand_spatial_points(
    points: list[HeatmapPoint],
    north: float,
    south: float,
    east: float,
    west: float,
    zoom: int,
    grid: int,
    max_total: int = 1500,
) -> list[HeatmapPoint]:
    """Duplicate each sample with neighbors so leaflet.heat blends smoothly."""
    if not points:
        return []

    step = spread_step_degrees(north, south, east, west, grid, zoom)
    expanded: list[HeatmapPoint] = []

    for p in points:
        for dlat_m, dlon_m, factor in _SPREAD_PATTERN:
            if len(expanded) >= max_total:
                return expanded
            intensity = min(1.0, max(0.0, p.intensity * factor))
            expanded.append(
                HeatmapPoint(
                    lat=p.lat + dlat_m * step,
                    lng=p.lng + dlon_m * step,
                    intensity=intensity,
                    value=p.value,
                )
            )
    return expanded


def focal_cluster(
    lat: float,
    lng: float,
    intensity: float,
    value: float | None,
    step: float,
    rings: int = 2,
) -> list[HeatmapPoint]:
    """Dense cluster around a selected location."""
    cluster: list[HeatmapPoint] = []
    for ring in range(1, rings + 2):
        ring_step = step * ring
        ring_factor = max(0.35, 1.0 - ring * 0.22)
        for dlat_m, dlon_m, factor in _SPREAD_PATTERN:
            if dlat_m == 0 and dlon_m == 0 and ring > 1:
                continue
            cluster.append(
                HeatmapPoint(
                    lat=lat + dlat_m * ring_step,
                    lng=lng + dlon_m * ring_step,
                    intensity=min(1.0, intensity * factor * ring_factor),
                    value=value,
                )
            )
    return cluster

File: frontend/src/lib/heatmap-spread.ts
import { clampIntensity } from "@/lib/heatmap-config";
import type { HeatmapPoint } from "@/types/intelligence";

const SPREAD_PATTERN: [number, number, number][] = [
  [0, 0, 1],
  [1, 0, 0.92],
  [-1, 0, 0.88],
  [0, 1, 0.85],
  [0, -1, 0.8],
  [1, 1, 0.78],
  [1, -1, 0.76],
  [-1, 1, 0.74],
  [-1, -1, 0.72],
  [2, 0, 0.65],
  [-2, 0, 0.62],
  [0, 2, 0.6],
  [0, -2, 0.58],
];

export function spreadStepForZoom(zoom: number, bounds?: {
  north: number;
  south: number;
  east: number;
  west: number;
}): number {
  if (bounds) {
    const latSpan = Math.max(0.01, bounds.north - bounds.south);
    const lonSpan = Math.max(0.01, bounds.east - bounds.west);
    const base = Math.max(latSpan, lonSpan) / 12;
    return base * (1 + Math.max(0, 14 - zoom) * 0.06);
  }
  return 0.012 * (1 + Math.max(0, 14 - zoom) * 0.05);
}

/**
 * Expand each API point with spatial neighbors for smooth GIS-style blending.
 */
export function densifyHeatmapPoints(
  points: HeatmapPoint[],
  zoom: number,
  bounds?: { north: number; south: number; east: number; west: number },
  maxTotal = 1200
): HeatmapPoint[] {
  if (points.length === 0) return [];

  const step = spreadStepForZoom(zoom, bounds);
  const out: HeatmapPoint[] = [];

  for (const p of points) {
    for (const [dlatM, dlonM, factor] of SPREAD_PATTERN) {
      if (out.length >= maxTotal) return out;
      out.push({
        lat: p.lat + dlatM * step,
        lng: p.lng + dlonM * step,
        intensity: clampIntensity(p.intensity * factor),
        value: p.value,
      });
    }
  }
  return out;
}

/**
 * Extra dense spread around the user's selected pin.
 */
export function generateFocalSpread(
  lat: number,
  lng: number,
  baseIntensity: number,
  zoom: number,
  value?: number | null
): HeatmapPoint[] {
  const step = spreadStepForZoom(zoom) * 1.1;
  const intensity = clampIntensity(baseIntensity);
  const cluster: HeatmapPoint[] = [];

  for (let ring = 0; ring < 3; ring++) {
    const ringStep = step * (ring + 1);
    const ringFactor = Math.max(0.4, 1 - ring * 0.2);
    for (const [dlatM, dlonM, factor] of SPREAD_PATTERN) {
      if (ring > 0 && dlatM === 0 && dlonM === 0) continue;
      cluster.push({
        lat: lat + dlatM * ringStep,
        lng: lng + dlonM * ringStep,
        intensity: clampIntensity(intensity * factor * ringFactor),
        value: value ?? undefined,
      });
    }
  }
  return cluster;
}

export function mergeHeatmapPoints(
  ...groups: HeatmapPoint[][]
): HeatmapPoint[] {
  const seen = new Set<string>();
  const merged: HeatmapPoint[] = [];
  for (const group of groups) {
    for (const p of group) {
      const key = `${p.lat.toFixed(5)}:${p.lng.toFixed(5)}`;
      if (seen.has(key)) continue;
      seen.add(key);
      merged.push(p);
    }
  }
  return merged;
}
